pennylane/labs/tests/trotter_error/product_formulas/test_bch.py

```diff
@@ -0,0 +1,293 @@
+
+    commutator_coeffs = {
+        ("X", "X", "X", "Y", "X"): 0.0047,
+        ("X", "X", "Y", "Y", "X"): 0.0057,
+        ("X", "Y", "X", "Y", "X"): 0.0046,
+        ("X", "Y", "Y", "Y", "X"): 0.0074,
+        ("Y", "X", "X", "Y", "X"): 0.0097,
+        ("Y", "X", "Y", "Y", "X"): 0.0097,
+        ("Y", "Y", "X", "Y", "X"): 0.0173,
+        ("Y", "Y", "Y", "Y", "X"): 0.0284,
+    }
+
+    upper_bound = 0
+    for comm, coeff in commutator_coeffs.items():
+        mat = nested_commutator([fragments[frag] for frag in comm])
+        upper_bound += coeff * np.linalg.norm(mat)
+
+    assert np.linalg.norm(fourth_order_approx - actual) <= (t**5) * upper_bound
+
+
+X = "X"
+Y = "Y"
+Z = "Z"
+
+
+@pytest.mark.parametrize(
+    "frag_labels, frag_coeffs, max_order, expected",
+    [
+        (
+            [X, Y],
+            [1, 1],
+            5,
+            [
+                {(X,): 1, (Y,): 1},
+                {(X, Y): 1 / 2},
+                {(X, X, Y): 1 / 12, (Y, X, Y): -1 / 12},
+                {(X, Y, X, Y): -1 / 24},
+                {
+                    (X, X, X, X, Y): -1 / 720,
+                    (X, X, Y, X, Y): -1 / 120,
+                    (X, Y, Y, X, Y): -1 / 360,
+                    (Y, X, X, X, Y): 1 / 360,
+                    (Y, X, Y, X, Y): 1 / 120,
+                    (Y, Y, Y, X, Y): 1 / 720,
+                },
+            ],
+        ),
+        # (
+        #    [X, Y, Z],
+        #    [1, 1, 1],
+        #    3,
+        #    [
+        #        {(X,): 1, (Y,): 1, (Z,): 1},
+        #        {
+        #            (X, Y): 1 / 2,
+        #            (X, Z): 1 / 2,
+        #            (Y, Z): 1 / 2,
+        #        },
+        #        {
+        #            (X, X, Y): 1 / 12,
+        #            (X, X, Z): 1 / 12,
+        #            (X, Y, Z): 1 / 3,
+        #            (Y, X, Y): -1 / 12,
+        #            (Y, X, Z): -1 / 6,
+        #            (Y, Y, Z): 1 / 12,
+        #            (Z, X, Z): -1 / 12,
+        #            (Z, Y, Z): -1 / 12,
+        #        },
+        #    ],
+        # ),
+        (
+            [X, Y, X],
+            [1, 1, 1],
+            3,
+            [
+                {(X,): 2, (Y,): 1},
+                {},
+                {(X, X, Y): -1 / 6, (Y, X, Y): -1 / 6},
+            ],
+        ),
+        (
+            [X, Y, X],
+            [1 / 2, 1, 1 / 2],
+            3,
+            [{(X,): 1, (Y,): 1}, {}, {(X, X, Y): -1 / 24, (Y, X, Y): -1 / 12}],
+        ),
+    ],
+)
+def test_bch_expansion(frag_labels, frag_coeffs, max_order, expected):
+    """Test against BCH expansion. The expected values come from Sections 4 and 5 of `arXiv:2006.15869 <https://arxiv.org/pdf/2006.15869>`"""
+
+    product_formula = ProductFormula(frag_labels, coeffs=frag_coeffs)
+    actual = bch_expansion(product_formula, max_order)
+
+    for i, order in enumerate(actual):
+        for comm in order:
+            assert np.isclose(order[comm], expected[i][comm])
```

Comment:
[non-blocking] Ideally, better not to put assert in a loop. It makes it hard to fix possible failures.